src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs

```diff
@@ -822,17 +822,87 @@ internal override TypeSymbol SetUnknownNullabilityForReferenceTypes()
                 }
             }
 
-            if (!haveChanges)
+            TypeSymbol result = this;
+            if (haveChanges)
             {
-                allTypeArguments.Free();
-                return this;
+                var definition = this.OriginalDefinition;
+                TypeMap substitution = new TypeMap(definition.GetAllTypeParameters(), allTypeArguments.ToImmutable());
+                result = substitution.SubstituteNamedType(definition);
             }
-            else
+
+            allTypeArguments.Free();
+            return result;
+        }
+
+        internal override (bool, TypeSymbol) TryMergeNullability(TypeSymbol other, VarianceKind variance)
+        {
+            Debug.Assert(this.Equals(other, TypeCompareKind.IgnoreDynamicAndTupleNames));
+
+            if (!IsGenericType)
             {
-                TypeMap substitution = new TypeMap(this.OriginalDefinition.GetAllTypeParameters(),
-                                                   allTypeArguments.ToImmutableAndFree());
+                return (true, this);
+            }
+
+            var definition = this.OriginalDefinition;
+            var allTypeParameters = definition.GetAllTypeParameters();
+            var allTypeArguments = ArrayBuilder<TypeSymbolWithAnnotations>.GetInstance();
+            GetAllTypeArgumentsNoUseSiteDiagnostics(allTypeArguments);
+            var otherTypeArguments = ArrayBuilder<TypeSymbolWithAnnotations>.GetInstance();
+            ((NamedTypeSymbol)other).GetAllTypeArgumentsNoUseSiteDiagnostics(otherTypeArguments);
+
+            try
+            {
+                bool haveChanges = false;
+                for (int i = 0; i < allTypeArguments.Count; i++)
+                {
+                    TypeSymbolWithAnnotations oldTypeArgument = allTypeArguments[i];
+                    VarianceKind typeArgumentVariance = GetTypeArgumentVariance(variance, allTypeParameters[i].Variance);
+                    (bool success, TypeSymbolWithAnnotations newTypeArgument) = oldTypeArgument.TryMergeNullability(otherTypeArguments[i], typeArgumentVariance);
+                    if (!success)
+                    {
+                        return (false, default);
+                    }
+                    if (!oldTypeArgument.IsSameAs(newTypeArgument))
+                    {
+                        allTypeArguments[i] = newTypeArgument;
+                        haveChanges = true;
+                    }
+                }
 
-                return substitution.SubstituteNamedType(this.OriginalDefinition);
+                TypeSymbol result = this;
+                if (haveChanges)
+                {
+                    TypeMap substitution = new TypeMap(allTypeParameters, allTypeArguments.ToImmutable());
+                    result = substitution.SubstituteNamedType(definition);
```

Comment:
>definition [](start = 62, length = 10)

This does not work for nested generic type within a generic type. The enclosing type might have to be merged too.

Reply:
Added a test.

---
In reply to: [224638987](https://github.com/dotnet/roslyn/pull/29727#discussion_r224638987) [](ancestors = 224638987)